### PR TITLE
feat: Add orm for Tools and clean up Tool logic

### PR DESCRIPTION
**Pull request by mattzh72**
## Description

This is a large refactor of how Tools are being instantiated today.

- We create a Tool ORM model and a corresponding Tool manager. 
- We change the Tool interface to require both user_id and organization_id on creation
- We integrate these changes across all the surfaces in the codebase (there are many) that use Tools

@norton120 again credits for inspiring the orm model for tool

## Testing

We write a suite of unit tests for the tool manager interface and also adapt existing tests to use our new interfaces.

### CODE REVIEW

File: examples/langchain_tool_usage.py [hunk 1]
```diff
@@ -33,7 +33,7 @@ def main():
     # We need to also import WikipediaAPIWrapper
     # The map is a mapping of the module name to the attribute name
     # langchain_community.utilities.WikipediaAPIWrapper
-    wikipedia_query_tool = Tool.from_langchain(
+    wikipedia_query_tool = ToolCreate.from_langchain(
```
Comment (sarahwooders) [inline examples/langchain_tool_usage.py:36]: why not just pass the langchain tool to `client.create_tool(...)`? I feel like using `ToolCreate` directly from the Python interface in this way is a bit weird.

Comment (sarahwooders) [reply examples/langchain_tool_usage.py:36]: Suggested API: 
```
tool = client.load_tool_from_langchain(langchain_tool, additional_imports_module_attr_map={"langchain_community.utilities": "WikipediaAPIWrapper"}
```

File: letta/client/client.py [hunk 2]
```diff
@@ -2288,16 +2253,14 @@ def update_tool(
         Returns:
             tool (Tool): Updated tool
         """
-        if func:
-            source_code = parse_source_code(func)
-        else:
-            source_code = None
-
-        source_type = "python"
+        update_data = {
```
Comment (sarahwooders) [inline letta/client/client.py:2256]: why not just pass this into `ToolUpdate`?

File: letta/config.py [hunk 3]
```diff
@@ -45,7 +45,7 @@ def set_field(config, section, field, value):
 @dataclass
 class LettaConfig:
     config_path: str = os.getenv("MEMGPT_CONFIG_PATH") or os.path.join(LETTA_DIR, "config")
-    anon_clientid: str = DEFAULT_USER_ID
+    anon_clientid: str = UserManager.DEFAULT_USER_ID
```
Comment (sarahwooders) [inline letta/config.py:48]: I think we should just delete `anon_clientid`, and keep user/org ID + name as constants

File: letta/client/client.py [hunk 4]
```diff
@@ -2188,48 +2188,8 @@ def delete_human(self, id: str):
...
-                    source_type=tool.source_type,
-                    source_code=tool.source_code,
-                    name=tool.name,
-                    json_schema=tool.json_schema,
-                    tags=tool.tags,
-                ),
-                user_id=self.user_id,
-                update=update,
-            )
+    def create_or_update_tool(self, tool_create: ToolCreate) -> Tool:
```
Comment (sarahwooders) [inline letta/client/client.py:2191]: maybe unpack the arguments here to avoid requiring a `ToolCreate`?

File: letta/client/client.py [hunk 5]
```diff
@@ -2327,19 +2290,20 @@ def delete_tool(self, id: str):
...
             name (str): Name of the tool
 
         Returns:
             id (str): ID of the tool (`None` if not found)
         """
-        return self.server.get_tool_id(name, self.user_id)
+        tool = self.server.tool_manager.get_tool_by_name_and_org_id(tool_name=name, organization_id=self.org_id)
+        return tool.id
 
     def tool_with_name_and_user_id_exists(self, tool: Tool) -> bool:
```
Comment (sarahwooders) [inline letta/client/client.py:2308]: What is this function?

File: examples/langchain_tool_usage.py (hunk 1 above)
Comment (mattzh72) [reply examples/langchain_tool_usage.py:36]: Ack

File: letta/client/client.py (hunk 4 above)
Comment (mattzh72) [reply letta/client/client.py:2191]: factored this out, deleted this exposed endpoint

File: letta/client/client.py (hunk 5 above)
Comment (mattzh72) [reply letta/client/client.py]: deleted

File: letta/config.py (hunk 3 above)
Comment (mattzh72) [reply letta/config.py:48]: i'm going to do this in a follow up PR

File: letta/client/client.py (hunk 2 above)
Comment (mattzh72) [reply letta/client/client.py:2282]: sorry this code is very hard to understand, let me refactor it - basically trying to only include keys that are non-null

File: letta/server/rest_api/admin/users.py [hunk 6]
```diff
@@ -25,18 +25,18 @@ def get_all_users(cursor: Optional[str] = Query(None), limit: Optional[int] = Qu
...
-        Create a new user in the database
-        """
-        try:
-            user = server.create_user(request)
-        except HTTPException:
-            raise
-        except Exception as e:
-            raise HTTPException(status_code=500, detail=f"{e}")
-        return user
+    # @router.post("/users", tags=["admin"], response_model=User)
```
Comment (sarahwooders) [inline letta/server/rest_api/admin/users.py:28]: add back?

File: letta/server/rest_api/routers/v1/tools.py [hunk 7]
```diff
@@ -50,26 +48,26 @@ def get_tool_id(
...
+    try:
+        tool = server.tool_manager.get_tool_by_name_and_org_id(tool_name=tool_name, organization_id=actor.organization_id)
+        return tool.id
+    except NoResultFound:
+        raise HTTPException(status_code=404, detail=f"Tool with name {tool_name} and organization id {actor.organization_id} not found.")
 
 
-@router.get("/", response_model=List[Tool], operation_id="list_tools")
-def list_all_tools(
+@router.get("/", response_model=List[Tool], operation_id="list_tools_for_org")
```
Comment (sarahwooders) [inline letta/server/rest_api/routers/v1/tools.py:58]: rename to "list_tools"

Comment (mattzh72) [reply letta/server/rest_api/routers/v1/tools.py:58]: done

File: letta/server/rest_api/admin/users.py (hunk 6 above)
Comment (mattzh72) [reply letta/server/rest_api/admin/users.py]: added

File: letta/server/server.py [hunk 8]
```diff
@@ -755,22 +756,6 @@ def run_command(self, user_id: str, agent_id: str, command: str) -> LettaUsageSt
                 command = command[1:]  # strip the prefix
         return self._command(user_id=user_id, agent_id=agent_id, command=command)
 
-    def create_user(self, request: UserCreate) -> User:
```
[the remark this answers is not in the document]
Comment (sarahwooders) [reply letta/server/server.py:758]: For any new org, we should also make sure we load the default tools in if that's what the settings specify.

[the remark this answers is not in the document]
Comment (mattzh72) [reply letta/server/server.py:758]: discussed offline, offloaded the logic to user manager

**Review by sarahwooders — APPROVED** [review]
lgtm!

**Pull request merged**